Civi/API/Api4SelectQuery.php

```diff
@@ -147,4 +158,148 @@ protected function getField($fieldName) {
+          throw new \Exception('Invalid dot notation in select');
+      }
+
+      if (!$customFieldData) {
+        continue;
+      }
+
+      $tableAlias = $customFieldData[0];
+      $columnName = $customFieldData[1];
+      $field = sprintf('%s.%s', $tableAlias, $columnName);
+
+      $this->selectFields[$field] = $selectAlias;
+    }
+  }
+
+  /**
+   * @param $customField
+   *   The field name in the format CustomGroupName.CustomFieldName
+   *
+   * @return array|null
+   *   An array containing the added table alias and column name
+   */
+  protected function addDotNotationCustomField($customField) {
+    $parts = explode('.', $customField);
+    $groupName = ArrayHelper::value(0, $parts);
+    $fieldName = ArrayHelper::value(1, $parts);
+
+    $customGroup = $this->getCustomGroup($this->entity, $groupName);
+    $tableName = ArrayHelper::value('table_name', $customGroup);
+    $customField = $this->getCustomField($customGroup['id'], $fieldName);
+    $columnName = ArrayHelper::value('column_name', $customField);
+
+    if (!$tableName || !$columnName) {
+      return NULL;
+    }
+
+    return $this->addCustomField(
+      array('table_name' => $tableName, 'column_name' => $columnName),
+      'INNER'
+    );
+  }
+
+  /**
+   * @param $field
+   *
+   * @return array
+   *   An array containing the option value table alias and column name
+   */
+  protected function addDotNotationCustomFieldWithOptionValue($field) {
+    $parts = explode('.', $field);
+    $groupName = ArrayHelper::value(0, $parts);
+    $fieldName = ArrayHelper::value(1, $parts);
+
+    $customGroup = $this->getCustomGroup($this->entity, $groupName);
+    $customField = $this->getCustomField($customGroup['id'], $fieldName);
+
+    $optionValueField = ArrayHelper::value(2, $parts);
+    $addedField = $this->addDotNotationCustomField($field);
+    $customValueAlias = $addedField[0];
+    $customValueColumn = $addedField[1];
+
+    $optionGroupID = ArrayHelper::value('option_group_id', $customField);
+
+    if (NULL === $optionGroupID) {
+      throw new \API_Exception(
+        'Cannot select option value field for a custom field that does not'
+        . ' have an option group defined'
+      );
+    }
+
+    $optionValueAlias = sprintf('%s_to_%s', self::MAIN_TABLE_ALIAS, 'option_value');
+    $optionValueMatching = sprintf(
+      '`%s`.value = `%s`.`%s`',
+      $optionValueAlias,
+      $customValueAlias,
+      $customValueColumn
+    );
+    $optionGroupRestriction = sprintf(
+      '`%s`.option_group_id =  %d',
+      $optionValueAlias,
+      $optionGroupID
+    );
+
+    $this->join(
+      'INNER',
+      'civicrm_option_value',
+      $optionValueAlias,
+      array($optionValueMatching, $optionGroupRestriction)
+    );
+
+    return array($optionValueAlias, $optionValueField);
+  }
+
+  /**
+   * @param string $extends
+   * @param string $groupName
+   *
+   * @return array|null
+   */
+  protected function getCustomGroup($extends, $groupName) {
```

Comment:
I'm curious to know why you need 2 params here. Pretty sure there's a unique constraint on custom group name, regardless of the entity it extends.